File: scripts/icon-search/render-svgs.ts
/* eslint-disable no-console */
/**
 * Render all SVG icons from @gravity-ui/icons to 224×224 PNG files.
 */

import fs from 'fs';
import path from 'path';

import sharp from 'sharp';

import {ICONS_LIBRARY_DIR, ICONS_LIBRARY_METADATA_PATH, RENDERED_ICONS_DIR} from './constants';

const TARGET_SIZE = 224;
const ICON_SIZE = 160;

type IconMeta = {
    name: string;
    svgName: string;
    componentName: string;
    style: string;
};

async function renderSvg(svgPath: string, outputPath: string): Promise<void> {
    const svgBuffer = fs.readFileSync(svgPath);
    const iconPng = await sharp(svgBuffer).resize(ICON_SIZE, ICON_SIZE).png().toBuffer();
    const offset = Math.floor((TARGET_SIZE - ICON_SIZE) / 2);

    await sharp({
        create: {
            width: TARGET_SIZE,
            height: TARGET_SIZE,
            channels: 3,
            background: {r: 255, g: 255, b: 255},
        },
    })
        .composite([{input: iconPng, left: offset, top: offset}])
        .png()
        .toFile(outputPath);
}

export async function renderSvgs() {
    if (!fs.existsSync(ICONS_LIBRARY_DIR)) {
        console.error(`Icons library directory not found: ${ICONS_LIBRARY_DIR}`);
        console.error('Make sure node_modules are installed (npm install).');
        process.exit(1);
    }

    fs.mkdirSync(RENDERED_ICONS_DIR, {recursive: true});

    const metadata = JSON.parse(fs.readFileSync(ICONS_LIBRARY_METADATA_PATH, 'utf-8'));
    const icons: IconMeta[] = metadata.icons;
    const total = icons.length;
    let rendered = 0;
    let skipped = 0;

    for (let i = 0; i < total; i++) {
        const icon = icons[i];
        const svgPath = path.join(ICONS_LIBRARY_DIR, 'svgs', `${icon.svgName}.svg`);

        if (!fs.existsSync(svgPath)) {
            console.log(`  SKIP: ${icon.svgName}.svg not found`);
            skipped++;
            continue;
        }

        const outputPath = path.join(RENDERED_ICONS_DIR, `${icon.svgName}.png`);
        await renderSvg(svgPath, outputPath);
        rendered++;

        if ((i + 1) % 100 === 0 || i + 1 === total) {
            console.log(`  [${i + 1}/${total}] rendered`);
        }
    }

    console.log(`\nDone: ${rendered} rendered, ${skipped} skipped.`);
    console.log(`Output: ${RENDERED_ICONS_DIR}`);
}

File: services/icon-search/scripts/render_svgs.py
"""Render all SVG icons from @gravity-ui/icons to 224x224 PNG files.

Requires cairosvg: pip install cairosvg
Run from services/icon-search/:
    python scripts/render_svgs.py
"""

import json
import sys
from pathlib import Path

import cairosvg
from PIL import Image
import io

REPO_ROOT = Path(__file__).resolve().parent.parent.parent.parent
SVGS_DIR = REPO_ROOT / "node_modules" / "@gravity-ui" / "icons" / "svgs"
METADATA_PATH = REPO_ROOT / "node_modules" / "@gravity-ui" / "icons" / "metadata.json"
OUTPUT_DIR = Path(__file__).resolve().parent.parent / "data" / "rendered_icons"

TARGET_SIZE = 224
ICON_SIZE = 160  # icon size within the 224x224 canvas (leaves padding)


def render_svg_to_png(svg_path: Path, output_path: Path) -> None:
    """Render an SVG to a 224x224 PNG with white background and centered icon."""
    png_bytes = cairosvg.svg2png(
        url=str(svg_path),
        output_width=ICON_SIZE,
        output_height=ICON_SIZE,
    )

    icon = Image.open(io.BytesIO(png_bytes)).convert("RGBA")

    canvas = Image.new("RGB", (TARGET_SIZE, TARGET_SIZE), (255, 255, 255))

    offset = (TARGET_SIZE - ICON_SIZE) // 2
    canvas.paste(icon, (offset, offset), mask=icon)

    canvas.save(output_path, "PNG")


def main():
    if not SVGS_DIR.exists():
        print(f"SVGs directory not found: {SVGS_DIR}")
        print("Make sure node_modules are installed (npm install).")
        sys.exit(1)

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    with open(METADATA_PATH) as f:
        metadata = json.load(f)

    icons = metadata["icons"]
    total = len(icons)
    rendered = 0
    skipped = 0

    for i, icon in enumerate(icons):
        svg_name = icon["svgName"]
        svg_path = SVGS_DIR / f"{svg_name}.svg"

        if not svg_path.exists():
            print(f"  SKIP: {svg_name}.svg not found")
            skipped += 1
            continue

        output_path = OUTPUT_DIR / f"{svg_name}.png"
        render_svg_to_png(svg_path, output_path)
        rendered += 1

        if (i + 1) % 100 == 0 or i + 1 == total:
            print(f"  [{i + 1}/{total}] rendered")

    print(f"\nDone: {rendered} rendered, {skipped} skipped.")
    print(f"Output: {OUTPUT_DIR}")


if __name__ == "__main__":
    main()
